voila/app.py

```diff
@@ -348,6 +346,25 @@ def setup_template_dirs(self):
                 self.static_paths,
                 self.template_paths,
                 self.voila_configuration.template)
+            # look for possible template-related config files
+            template_conf_dir = [os.path.dirname(k) for k in self.nbconvert_template_paths]
+            conf_paths = [os.path.join(d, 'conf.json') for d in template_conf_dir]
+            for p in conf_paths:
+                # see if config path corresponds to template in use
+                if os.path.basename(os.path.dirname(p)) == self.voila_configuration.template:
+                    # see if config file exists
+                    if os.path.exists(p):
+                        # load the template-related config
+                        loader = JSONFileConfigLoader('conf.json', template_conf_dir)
+                        conf = loader.load_config()
+                        # check that config file is meant for template in use
+                        assert conf.traitlet_configuration['base_template'] == self.voila_configuration.template
+                        # create sub-config as Config instance for merging (see next)
+                        conf.VoilaConfiguration = Config(conf.traitlet_configuration)
+                        # merge config from file with otherwise voila config, ensuring CLI config priority
+                        conf.merge(self.voila_configuration.config)
+                        # pass merged config to overall voila config
```

Comment:
I think instead of relying on the JSONFileConfigLoader, it would be easier to load the json using the json library, use the merge functionality of jupyter_server, and pass the resulting merged dict to the Config constructor. What do you think?

Reply:
Right, it would be more straightforward, more readable to use plain JSON... And then we wouldn't have to sacrifice this precious recursive merge. You have convinced me!

Reply:
Addressed in 52f49db618347494846b21e16047d30aa935f36b (reminiscent #326 and sparing a couple lines of code).